snuba/optimize.py

```diff
@@ -1,20 +1,46 @@
-from datetime import timedelta
-from typing import Sequence
 import logging
+from datetime import datetime, timedelta
+from typing import Optional, Sequence
 
 from snuba import util
+from snuba.clickhouse.native import ClickhousePool
+from snuba.datasets.schemas.tables import TableSchema
+from snuba.datasets.storages import StorageKey
+from snuba.datasets.storages.factory import get_storage
 
 logger = logging.getLogger("snuba.optimize")
 
 
-def run_optimize(clickhouse, database, table, before=None):
-    parts = get_partitions_to_optimize(clickhouse, database, table, before)
-    optimize_partitions(clickhouse, database, table, parts)
+STORAGE_PARTITION_KEYS = {
+    StorageKey.EVENTS: [util.PartSegment.DATE, util.PartSegment.RETENTION_DAYS],
+    StorageKey.ERRORS: [util.PartSegment.RETENTION_DAYS, util.PartSegment.DATE],
+    StorageKey.TRANSACTIONS: [util.PartSegment.RETENTION_DAYS, util.PartSegment.DATE],
+}
+
+
+def run_optimize(
+    clickhouse: ClickhousePool,
+    storage_key: StorageKey,
```

Comment:
Could you please pass in the storage instead of the storage key? You can get the storage key from the storage (weirdly enough this is not in the abstract Storage class but only in the readable subclass, which seems wrong and easy to fix).

Keeping the access to the global storage factory only as close as possible to the entry points (where we turn a string into a storage, thus it is inevitable) removes such dependency from this function thus making it easier to test.
This is because it would depend only on the input it receives, thus there would not be a need to mock or tweak the factory to unit test this class in a way that is totally independent on the data model.